lib/src/revset.rs

```diff
@@ -1269,8 +1272,51 @@ fn parse_function_argument_to_string_pattern(
     pair: Pair<Rule>,
     state: ParseState,
 ) -> Result<StringPattern, RevsetParseError> {
-    let needle = parse_function_argument_as_literal("string", name, pair, state)?;
-    Ok(StringPattern::Substring(needle))
+    let span = pair.as_span();
+    let make_error = |message| {
+        RevsetParseError::with_span(
+            RevsetParseErrorKind::InvalidFunctionArguments {
+                name: name.to_string(),
+                message,
+            },
+            span,
+        )
+    };
+    let make_type_error = || make_error("Expected function argument of string pattern".to_owned());
+    let expression = parse_expression_rule(pair.into_inner(), state)?;
+    let pattern = match expression.as_ref() {
+        RevsetExpression::CommitRef(RevsetCommitRef::Symbol(symbol)) => {
+            let needle = symbol.to_owned();
+            StringPattern::Substring(needle)
+        }
+        // TODO: Add proper parsed node if we drop support for legacy x:y range
+        RevsetExpression::DagRange {
+            roots,
+            heads,
+            is_legacy: true,
+        } => {
+            // TODO: quoted string shouldn't be allowed as a pattern kind
+            let RevsetExpression::CommitRef(RevsetCommitRef::Symbol(kind)) = roots.as_ref() else {
+                return Err(make_type_error());
+            };
+            let RevsetExpression::CommitRef(RevsetCommitRef::Symbol(needle)) = heads.as_ref()
+            else {
+                return Err(make_type_error());
+            };
+            match kind.as_ref() {
+                "literal" => StringPattern::Literal(needle.clone()),
+                // TODO: maybe add explicit kind for substring match?
```

Comment:
Yes, that would seem consistent. We can also reevaluate if we want the default to be substring matching.